### PR TITLE
feat(babel): pass rollup context as this context into override config function

### PR DESCRIPTION
**Pull request by KingSora**
## Rollup Plugin Name: `babel`

This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

If yes, then include "BREAKING CHANGES:" in the first commit message body, followed by a description of what is breaking. 

List any relevant issue numbers:

### Description

Implements my suggestion from https://github.com/rollup/plugins/issues/774
Basically its now possible that you can pass information from the rollup context (like `getCombinedSourcemap()`) into any babel plugin with the `createBabelInputPluginFactory`, because the rollup context is now passed as `this` context into the `override config` function. (like its done in the `override result` function).

Also updated `index.d.ts` for typescript users.

### CODE REVIEW

**Review by shellscape — APPROVED** [review]

**Comment by shellscape** [comment]
@Andarist any thoughts on this one?

File: packages/babel/test/as-input-plugin.js [hunk 1]
```diff
@@ -377,6 +377,7 @@ test('supports customizing the loader', async (t) => {
   const customBabelPlugin = createBabelInputPluginFactory(() => {
     return {
       config(cfg) {
+        t.true(typeof this === 'object', 'override config this context is rollup context');
```
Comment (Andarist) [inline packages/babel/test/as-input-plugin.js:380]: could u test smth more distinct? it's rather easy to regress based on this simple check

File: packages/babel/types/index.d.ts [hunk 2]
```diff
@@ -53,10 +53,12 @@ export type RollupBabelCustomOutputPluginOptions = (
   pluginOptions: RollupBabelOutputPluginOptions;
 };
 export type RollupBabelCustomPluginConfig = (
+  this: PluginContext | TransformPluginContext,
```
Comment (Andarist) [inline packages/babel/types/index.d.ts:56]: I don't think this should be a union - this only gets called in a single place in the code and thus the type should be always the same. Since it's called from within a `transform` hook I would suspect that this can be simplified to:
```suggestion
  this: TransformPluginContext,
```

**Comment by Andarist** [comment]
@shellscape looks good overall, I've left just some minor comments - I would actually treat this as a bug fix (this was always the intention, it has regressed during some refactors).

File: packages/babel/types/index.d.ts (hunk 2 above)
Comment (KingSora) [reply packages/babel/types/index.d.ts:56]: I did this because the `createBabelOutputPluginFactory` calls the `transformCode` function in the `renderChunk` function where rollup passes only the `PluginContext`. 
Only the `createBabelInputPluginFactory` calls `transformCode` in the `transform` function where rollup passes the `TransformPluginContext`.

File: packages/babel/test/as-input-plugin.js (hunk 1 above)
Comment (KingSora) [reply packages/babel/test/as-input-plugin.js:380]: @Andarist do you have an idea how to do that? Unfortunately I'm no expert in rollup, so I don't know wether I can do something like `instanceof` or similar.

Comment (Andarist) [reply packages/babel/test/as-input-plugin.js:380]: just maybe try to check some example property/methods that exist on this object?

File: packages/babel/types/index.d.ts (hunk 2 above)
Comment (Andarist) [reply packages/babel/types/index.d.ts:56]: Good catch. It seems though that we should not reuse the `RollupBabelCustomPluginResult` and `RollupBabelCustomPluginConfig` types for both `RollupBabelCustomInputPlugin` and `RollupBabelCustomOutputPlugin`. I think we should split those so the `this` could be typed correctly for both cases

File: packages/babel/test/as-input-plugin.js (hunk 1 above)
Comment (KingSora) [reply packages/babel/test/as-input-plugin.js:381]: @Andarist alright, done.. Wanted to do this initially, but then I thought such checks could also be a bit shaky.
3b0e78b

File: packages/babel/types/index.d.ts (hunk 2 above)
Comment (KingSora) [reply packages/babel/types/index.d.ts]: @Andarist Changed things up a bit in: 3b0e78b Please take a look

**Review by Andarist — APPROVED** [review]

**Comment by shellscape** [comment]
@Andarist I kept this marked as a feat because from a high level view it really looks like more of a feature that contains a bugfix (it happens). On the version graph, there's not much difference there so I think we're safe. If you disagree with that, I can always force the commit back to a `fix`
